File: MyProject/Source/MyProject/Private/BasicFrame/Libs/Task/TaskThread.cpp
#include "MyProject.h"
#include "TaskThread.h"

TaskThread::TaskThread(std::string name, TaskQueue taskQueue)
    : base(null, null)
{
    mTaskQueue = taskQueue;
    mCondition = new MCondition(name);
}

/**
 *brief 线程回调函数
 */
void TaskThread::threadHandle()
{
    while (!mIsExitFlag)
    {
        mCurTask = mTaskQueue.pop();
        if(mCurTask != default(ITask))
        {
            mCurTask.runTask();
        }
        else
        {
            mCondition.wait();
        }
    }
}

bool TaskThread::notifySelf()
{
    if(mCondition.canEnterWait)
    {
        mCondition.notifyAll();
        return true;
    }

    return false;
}

File: Project/MyProject/Tool/CSToolKit/Project/ToolKit/ToolKit/Source/Libs/Task/TaskThread.cs
namespace SDK.Lib
{
    /**
     * @brief 任务线程
     */
    public class TaskThread : MThread
    {
        protected TaskQueue mTaskQueue;
        protected MCondition mCondition;
        protected ITask mCurTask;

        public TaskThread(string name, TaskQueue taskQueue)
            : base(null, null)
        {
            this.mTaskQueue = taskQueue;
            this.mCondition = new MCondition(name);
        }

        /**
         *brief 线程回调函数
         */
        override public void threadHandle()
        {
            while (!this.mIsExitFlag)
            {
                this.mCurTask = this.mTaskQueue.pop();

                if(this.mCurTask != default(ITask))
                {
                    this.mCurTask.runTask();
                }
                else
                {
                    this.mCondition.wait();
                }
            }
        }

        public bool notifySelf()
        {
            if(this.mCondition.getCanEnterWait())
            {
                this.mCondition.notifyAll();
                return true;
            }

            return false;
        }
    }
}